engine-src/Game/LambdaHack/Client/UI/SessionUI.hs

```diff
@@ -54,7 +54,9 @@ data SessionUI = SessionUI
                                     -- ^ parameters of the current run, if any
   , shistory       :: History       -- ^ history of messages
   , spointer       :: K.PointUI     -- ^ mouse pointer position
-  , slastRecord    :: LastRecord    -- ^ state of key sequence recording
+  , slastAction    :: Maybe K.KM    -- ^ last pressed key
+  , smacroBuffer   :: [K.KM]        -- ^ in-game recorded macro
```

Comment:
It's question of style, but I'd make `smacroBuffer :: Either [K.KM] [K.KM]`. Then my tip to eliminate the last pressed key of a macro when running the macro doesn't make sense (we "read" smacroBuffer anyway) and perhaps even `reverse` can be ASAP, for a silly optimization that will never matter.

`Either [K.KM] [K.KM]` is not satisfying, because both types are the same, so we don't gain any type safety. Which is why say it's personal style and completely up to you.

Reply:
To be clear --- then we can remove `srecording`.

Reply:
Sure thing, that's a neat idea. Tomorrow I'll look closer how it works up. Enough adventures for today :-)